soccer/planning/Trajectory.cpp

```diff
@@ -0,0 +1,344 @@
+    while (cursor.has_value() && cursor.value().stamp < actual_end) {
+        result_instants.push_back(cursor.value());
+        cursor.next_knot();
+    }
+
+    // The above calculation will always miss the last instant.
+    result_instants.push_back(Trajectory::interpolatedInstant(
+        result_instants.back(), cursor.value(), actual_end));
+
+    return Trajectory{std::move(result_instants)};
+}
+
+Trajectory::Cursor Trajectory::cursor(RJ::Time start_time) const {
+    return Cursor{*this, start_time};
+}
+
+Trajectory::Cursor Trajectory::cursor_begin() const {
+    return Cursor{*this, instants_.begin()};
+}
+
+void Trajectory::draw(
+    DebugDrawer* drawer,
+    std::optional<Geometry2d::Point> alt_text_position) const {
+    if (instants_.size() > 1) {
+        Packet::DebugRobotPath* dbgPath = drawer->addDebugPath();
+        dbgPath->set_layer(drawer->findDebugLayer("Motion"));
+
+        for (const RobotInstant& instant : instants_) {
+            Packet::DebugRobotPath::DebugRobotPathPoint* pt =
+                dbgPath->add_points();
+            *pt->mutable_pos() = instant.pose.position();
+            *pt->mutable_vel() = instant.velocity.linear();
+        }
+    }
+
+    if (debug_text_) {
+        Geometry2d::Point textPos;
+
+        // Only use the backup position if there's no trajectory.
+        if (!empty()) {
+            textPos = first().pose.position() + Geometry2d::Point(0.1, 0);
+        } else if (alt_text_position.has_value()) {
+            textPos = alt_text_position.value();
+        } else {
+            return;
+        }
+
+        drawer->drawText(QString::fromStdString(debug_text_.value()), textPos,
+                         QColor(100, 100, 255, 100), "PlanningDebugText");
+    }
+}
+
+bool Trajectory::nearly_equal(const Trajectory& a, const Trajectory& b,
+                              double tolerance) {
+    auto a_it = a.instants_begin();
+    auto b_it = b.instants_begin();
+    for (; a_it != a.instants_end() && b_it != b.instants_end();
+         ++a_it, ++b_it) {
+        if (!RobotInstant::nearly_equals(*a_it, *b_it, tolerance)) {
+            return false;
+        }
+    }
+
+    // If there is more time to go in either trajectory, they are different.
+    // Otherwise, they are identical.
+    return a_it == a.instants_end() && b_it == b.instants_end();
+}
+
+Trajectory::Cursor::Cursor(const Trajectory& trajectory,
+                           RobotInstantSequence::const_iterator iterator)
+    : trajectory_{trajectory}, iterator_{iterator}, time_{iterator->stamp} {}
+
+Trajectory::Cursor::Cursor(const Trajectory& trajectory)
+    : Cursor{trajectory, trajectory.instants_.begin()} {}
+
+Trajectory::Cursor::Cursor(const Trajectory& trajectory, RJ::Time start_time)
+    : Cursor{trajectory} {
+    seek(start_time);
+}
+
+RobotInstant Trajectory::Cursor::value() const {
+    RobotInstant instant_0 = *iterator_;
+    if (instant_0.stamp == time_) {
```

Comment:
`*iterator_.stamp` and `time_` are only equal if it's the last instant in the vector right?

Reply:
Yes. This check is necessary to avoid dereferencing the `end()` iterator in the case that we're pointing to the end.